File: edition4/js/ch01/1-2.js
/**
 * Reverses a c style string
 *
 * NOTE: using an array to represent a c style string, since string are
 * immutable in javascript.
 */
function reverseString(str) {
  var head = 0;
  var tail = 0;

  while (str[tail] !== '\0') {
    tail += 1;
  }
  tail -= 1;

  while (head < tail) {
    var temp = str[head];
    str[head] = str[tail];
    str[tail] = temp;

    head += 1;
    tail -= 1;
  }

  return str;
}

// tests
var assert = require('assert');
var mocha = require('mocha');

describe('reverseString', function() {

  it('should succeed', function() {
    var str = 'This is a test string\0';
    var actual = reverseString(str.split(''));
    var expected = 'gnirts tset a si sihT\0';
    assert.equal(actual.join(''), expected);
  });

});

File: edition4/c/ch01/1-2.c
#include <assert.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

void reverseString(char *str) {
  char *end = str;
  char temp;
  
  if (str) {
    // find end of string
    while(*end) {
      ++end;
    }
    --end;

    while(str < end) {
      // swap chars
      temp = *str;
      *str = *end;
      *end = temp;

      // move to next spot
      ++str;
      --end;
    }
  }
}

int main() {
  // test
  char str1[] = "This is a test string";
  reverseString(str1);
  assert(strcmp(str1, "gnirts tset a si sihT") == 0);
  
  return 0;
}
